devito/mpi/routines.py

```diff
@@ -298,48 +299,42 @@ def _make_all(self, f, hse, msg):
         return haloupdate, halowait
 
     def _make_copy(self, f, hse, key, swap=False):
-        buf_dims = []
-        buf_indices = []
-        for d in f.dimensions:
-            if d not in hse.loc_indices:
-                buf_dims.append(Dimension(name='buf_%s' % d.root))
-                buf_indices.append(d.root)
-        buf = Array(name=self.sregistry.make_name(prefix='buf'),
-                    dimensions=buf_dims, dtype=f.dtype, padding=0)
+        dims = [d.root for d in f.dimensions if d not in hse.loc_indices]
+        buf = Array(name='buf%s' % key, dimensions=dims, dtype=f.dtype, padding=0)
 
         f_offsets = []
         f_indices = []
         for d in f.dimensions:
-            offset = Symbol(name='o%s' % d.root)
+            offset = Symbol(name='o%s' % d.root, is_const=True)
             f_offsets.append(offset)
             f_indices.append(offset + (d.root if d not in hse.loc_indices else 0))
 
         if swap is False:
-            eq = DummyEq(buf[buf_indices], f[f_indices])
+            eq = Eq(buf[dims], f[f_indices])
             name = 'gather%s' % key
         else:
-            eq = DummyEq(f[f_indices], buf[buf_indices])
+            eq = Eq(f[f_indices], buf[dims])
             name = 'scatter%s' % key
 
-        iet = Expression(eq)
-        for i, d in reversed(list(zip(buf_indices, buf_dims))):
-            # The -1 below is because an Iteration, by default, generates <=
-            iet = Iteration(iet, i, d.symbolic_size - 1, properties=(PARALLEL, AFFINE))
+        eqns = []
+        eqns.extend([Eq(d.symbolic_min, 0) for d in dims])
+        eqns.extend([Eq(d.symbolic_max, d.symbolic_size - 1) for d in dims])
+        eqns.append(eq)
+
+        irs, _ = self._lower(eqns, **self._kwargs)
```

Comment:
@mloubout @georgebisbas this PR ships recursive compilation :)

I'm basically turning a list of eqns into a piece of IET, instead of constructing the IET directly

I'm only using the feature here ATM, but there probably are other places